src/modules/bulletproofs/rangeproof_uncompressed_impl.h

```diff
@@ -0,0 +1,152 @@
+/**********************************************************************
+ * Copyright (c) 2020 Andrew Poelstra                                 *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef _SECP256K1_MODULE_BULLETPROOFS_RP_UNCOMPRESSED_
+#define _SECP256K1_MODULE_BULLETPROOFS_RP_UNCOMPRESSED_
+
+#include "group.h"
+#include "scalar.h"
+
+#include "modules/bulletproofs/bulletproofs_util.h"
+
+/* Prover context data:
+ *    bytes 0-32:   x (hash challenge)
+ *    bytes 32-64:  y (hash challenge)
+ *    bytes 64-96:  z (hash challenge)
+ *    bytes 96-160: lr_generator data
+ */
+
+/* Generators (paper -> implementation):
+ *    g -> asset_gen
+ *    h -> default secp256k1 generator
+ *    *g*, *h* vectors -> gens
+ */
+
+/* Step 0 of the proof.
+ * Uses the value but not the blinding factor. Takes the complete commitment,
+ * but only to put it into the hash.
+ * Outputs the points A and S, encoded in 65 bytes (one byte with A's parity
+ * in the LSB, S's parity in the second-LSB, 32-byte A.x, 32-byte S.x).
+ * Updates the state to include the hashes y and z.
+ */
+static int secp256k1_bulletproofs_rangeproof_uncompressed_prove_step0_impl(
+    const secp256k1_ecmult_gen_context* ecmult_gen_ctx,
+    secp256k1_bulletproofs_prover_context* prover_ctx,
+    unsigned char* output,
+    const size_t n_bits,
+    const uint64_t value,
+    const uint64_t min_value,
+    const secp256k1_ge* commitp,
+    const secp256k1_ge* asset_genp,
+    const secp256k1_bulletproofs_generators* gens,
+    const unsigned char* nonce,
+    const secp256k1_scalar* enc_data,
+    const unsigned char* extra_commit,
+    size_t extra_commit_len
+) {
+    secp256k1_sha256 sha256;
+    unsigned char commit[32];
+    secp256k1_scalar alpha, rho;
+    secp256k1_scalar tmp_l, tmp_r;
+    secp256k1_gej gej;
+    secp256k1_ge ge;
+    size_t i;
+    int overflow;
+
+    memset(prover_ctx->data, 0, sizeof(prover_ctx->data));
+    memset(output, 0, 65);
```

Comment:
In 05e1d6cf8e6102e2ee86522b09bac1d9f4be7671:
ultra minor nit: 
Doing these memsets seems unnecessary as we are always writing data first before reading. But it's a good habit to follow in general, these are not done in steps 2 and 3.

Reply:
I think I'll leave this for a followup. It's a fair bit of work to audit 0-memsets for whether they're no-ops and/or whether they have some hypothetical security justification, and in the end the compiler will just do what it wants to.